### ISSUE TITLE
Bug: Pasting text immediately after an inline element (e.g. link) causes the text and its following siblings to be moved into the inline element

### ISSUE THREAD

**Issue by birtles** [facebook/lexical#4295]
Lexical version: 0.9.1

## Steps To Reproduce

1. Go to [Lexical Playground](https://playground.lexical.dev/)
2. Select some text, e.g. the word "playground"
3. Copy it to the clipboard
4. Position the cursor after the "Lexical website" link
5. Press Paste

## The current behavior

The pasted word ("playground") _and_ all the following text on that line end up being link-ified

https://user-images.githubusercontent.com/1232595/230831265-67ba23ab-c2fd-4767-ac23-93b56c5f762b.mp4

## The expected behavior

The pasted word should _not_ be link-ified and, at very least, the following text should not be link-ified either.

Note that this behavior reproduces with a non-collapsed selection too, provided the selection starts/ends at the end of an inline element (since for a non-collapsed selection we'll end up dropping the selection contents through a call to `removeText()`).

## Analysis

A bit of debugging points to the code here:

https://github.com/facebook/lexical/blob/933ec6ada2b2f25db807cb125de75952ea0f7fd3/packages/lexical/src/LexicalSelection.ts#L1409-L1412

Specifically, if the selection is at the start of a text run (anchor offset is zero) then we choose the previous sibling as the target and append the pasted text and its siblings into the previous sibling.

I'm not sure what the reasoning is for this logic but it produces surprising results for this particular case.

**Comment by shirakaba** [comment]
I've put together some test cases for this (all failing ones are marked with a comment) here: https://github.com/shirakaba/lexical/tree/insertnodes-after-inline-element

I'll try to come up with a fix to serve as a strawman.

**Comment by shirakaba** [comment]
I've pushed a strawman fix to that branch now. Just need to sort out the CLA to open a PR.

**Comment by shirakaba** [comment]
Adding some extra context emerging from @zurfyx's discussion in #4346:

> Additionally, prior to this [strawman] fix, it also seems like the text after is getting merged into the previous target element. It sems like the root cause is deeper than this.
> 
> Can we check why this is the case?

This seems to be due to this line here, where the siblings (the nodes to the right of the anchor node at the time of insertion) are handled:

https://github.com/facebook/lexical/blob/933ec6ada2b2f25db807cb125de75952ea0f7fd3/packages/lexical/src/LexicalSelection.ts#L1675

The intention (according to the code comment at the declaration of `nextSiblings`) is to "append them after the last node we're inserting", and in this case they're indeed being appended after the inserted node (which is unfortunately in the wrong place, inside the inline element).

I think in this case we want to insert the siblings after the target (the inline element), so maybe we want to get onto this conditional branch instead:

https://github.com/facebook/lexical/blob/933ec6ada2b2f25db807cb125de75952ea0f7fd3/packages/lexical/src/LexicalSelection.ts#L1696

Probably we could do that by augmenting the conditional check that we do here:

https://github.com/facebook/lexical/blob/933ec6ada2b2f25db807cb125de75952ea0f7fd3/packages/lexical/src/LexicalSelection.ts#L1666

... in the same way that we end up doing here (where this discussion began):

https://github.com/facebook/lexical/blob/933ec6ada2b2f25db807cb125de75952ea0f7fd3/packages/lexical/src/LexicalSelection.ts#L1409-L1412

In other words, if in the first place we change the check from just `$isElementNode(target)` to `$isElementNode(target) && !target.isInline()`, then we should make that same change in the second place, too.

**Comment by shirakaba** [comment]
Of note: I was curious why `insertText` works while `insertNodes` doesn't, and it occurred to me that it might be beneficial if the two were refactored to share the same code (after an initial conversion step), if possible. I was trying to compare the two, but unfortunately they share little similarity at present.

**Comment by birtles** [comment]
@zurfyx Do you want to re-open this? I think #4346 only added tests.

**Comment by evanmwillhite** [comment]
I am facing a similar issue with the code below. `linkCopy`, `textNode` and `newLink` all show the correct text selection but the link wraps the rest of the inline text no matter what.
```jsx
const linkNode = $createLinkNode(sanitizeUrl(linkUrl));
const textNode = $createTextNode(linkCopy);
const newLink = linkNode.append(textNode);
$insertNodes([newLink]);
```

https://github.com/facebook/lexical/assets/18293479/d89bd162-333f-428e-b031-44e854e15f0b
